src/irmin/type/type_core.ml

```diff
@@ -14,6 +14,19 @@
  * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
  *)
 
+exception Not_utf8
+
+let check_valid_utf8 str =
+  Uutf.String.fold_utf_8
+    (fun _ _ -> function `Malformed _ -> raise Not_utf8 | _ -> ())
```

Comment:
`Invalid_arg` probably suffices here (and for invalid combinator use generally), since they almost certainly correspond to programmer error rather than runtime exceptions.

Reply:
Alright, I'll switch to `Invalid_arg`.